From 175897b79ce33125619d0cb0263dde749f982ca5 Mon Sep 17 00:00:00 2001
From: locus84 <locus84@andestech.com>
Date: Thu, 9 Jan 2025 11:09:45 +0800
Subject: [PATCH 149/168] andes: spi: refinement of the data merge mode
 framework (#239)

1. Notify the user of a warning message when data_len is not 0x7.
2. Automatically disable data merge mode when the transmission length
   is not 4 byte aligned.

Signed-off-by: locus84 <locus84@andestech.com>

Reviewed-on: https://gitea.andestech.com/RD-SW/linux/pulls/239
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: CL Chin-Long Wang <cl634@andestech.com>
Co-authored-by: locus84 <locus84@andestech.com>
Co-committed-by: locus84 <locus84@andestech.com>
---
 drivers/spi/spi-atcspi200-dma.c         | 32 ++++++------
 drivers/spi/spi-atcspi200.c             | 67 +++++++++++++------------
 include/linux/soc/andes/spi-atcspi200.h | 27 +++++-----
 3 files changed, 64 insertions(+), 62 deletions(-)

diff --git a/drivers/spi/spi-atcspi200-dma.c b/drivers/spi/spi-atcspi200-dma.c
index 7b059aab9a56..ef081bfd0655 100644
--- a/drivers/spi/spi-atcspi200-dma.c
+++ b/drivers/spi/spi-atcspi200-dma.c
@@ -64,13 +64,13 @@ static inline int spi_dma_config_rx(struct atcspi200_spi *spi)
 	rxconf.src_addr = spi->dma_addr;
 	rxconf.dst_maxburst = 2;
 	rxconf.src_maxburst = 2;
-#ifdef	CONFIG_SPI_ATCSPI200_DATA_MERGE
-	rxconf.src_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
-	rxconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
-#else
-	rxconf.src_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
-	rxconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
-#endif
+	if (spi->data_merge) {
+		rxconf.src_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+		rxconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+	} else {
+		rxconf.src_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
+		rxconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
+	}
 
 	return dmaengine_slave_config(spi->rxchan, &rxconf);
 }
@@ -83,13 +83,13 @@ static inline int spi_dma_config_tx(struct atcspi200_spi *spi)
 	txconf.dst_addr = spi->dma_addr;
 	txconf.dst_maxburst = 2;
 	txconf.src_maxburst = 2;
-#ifdef	CONFIG_SPI_ATCSPI200_DATA_MERGE
-	txconf.src_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
-	txconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
-#else
-	txconf.src_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
-	txconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
-#endif
+	if (spi->data_merge) {
+		txconf.src_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+		txconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+	} else {
+		txconf.src_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
+		txconf.dst_addr_width = DMA_SLAVE_BUSWIDTH_1_BYTE;
+	}
 
 	return dmaengine_slave_config(spi->txchan, &txconf);
 }
@@ -101,7 +101,7 @@ static void atcspi200_spi_dma_callback(void *arg)
 	complete(dma_completion);
 }
 
-int atcspi200_spi_dma_setup(struct atcspi200_spi *spi, struct spi_mem_op *op)
+int atcspi200_spi_dma_setup(struct atcspi200_spi *spi, const struct spi_mem_op *op)
 {
 	int ret;
 	struct dma_async_tx_descriptor *desc;
@@ -152,7 +152,7 @@ int atcspi200_spi_dma_setup(struct atcspi200_spi *spi, struct spi_mem_op *op)
 	return ret;
 }
 
-int atcspi200_spi_dma_transfer(struct atcspi200_spi *spi, struct spi_mem_op *op)
+int atcspi200_spi_dma_transfer(struct atcspi200_spi *spi, const struct spi_mem_op *op)
 {
 	struct dma_chan *dma_ch;
 	int ret, timeout = DMA_DEV_TIMEOUT_MS;
diff --git a/drivers/spi/spi-atcspi200.c b/drivers/spi/spi-atcspi200.c
index 106e1a6884e2..32672e2b9a01 100644
--- a/drivers/spi/spi-atcspi200.c
+++ b/drivers/spi/spi-atcspi200.c
@@ -22,6 +22,8 @@
 #include <linux/minmax.h>
 #include <linux/soc/andes/spi-atcspi200.h>
 
+static bool ts_enable;
+
 static inline void atcspi200_spi_write(struct atcspi200_spi *spi, int offset, u32 value)
 {
 	iowrite32(value, spi->regs + offset);
@@ -60,9 +62,13 @@ static int spi_nor_setup(struct atcspi200_spi *spi)
 		if (!spi->timeout)
 			return -EINVAL;
 	}
-
 	spi->cmd_len = 0;
-	format_val = (DATA_LENGTH(8) | ADDR_LENGTH(3) | DATA_MERGE);
+
+	format_val =  DATA_LENGTH(DATA_BIT) | ADDR_LENGTH(ADDR_BIT);
+	spi->data_merge =
+		IS_ENABLED(CONFIG_SPI_ATCSPI200_DATA_MERGE) && (DATA_BIT == 8);
+	if (spi->data_merge)
+		format_val |= ATCSPI200_TRANSFMT_DATA_MERGE_MASK;
 	format_val |= ATCSPI200_TRANSFMT_CPHA_MASK;
 	format_val |= ATCSPI200_TRANSFMT_CPOL_MASK;
 	atcspi200_spi_write(spi, SPI_TRANSFMT, format_val);
@@ -148,39 +154,24 @@ static int atcspi200_spi_start(struct atcspi200_spi *spi, struct spi_transfer *t
 
 static inline void atcspi200_spi_tx(struct atcspi200_spi *spi, const void *dout)
 {
-#ifdef	CONFIG_SPI_ATCSPI200_DATA_MERGE
-	atcspi200_spi_write(spi, SPI_DATA, *(u32 *)dout);
-#else
-	atcspi200_spi_write(spi, SPI_DATA, *(u8 *)dout);
-#endif
+	if (spi->data_merge)
+		atcspi200_spi_write(spi, SPI_DATA, *(u32 *)dout);
+	else
+		atcspi200_spi_write(spi, SPI_DATA, *(u8 *)dout);
 }
 
 static inline int atcspi200_spi_rx(struct atcspi200_spi *spi, void *din, unsigned int bytes)
 {
-#ifdef	CONFIG_SPI_ATCSPI200_DATA_MERGE
-	u32 tmp_data;
-
-	tmp_data = atcspi200_spi_read(spi, SPI_DATA);
-	*(u32 *)din = tmp_data;
-#else
-	u8 tmp_data;
-
-	tmp_data = atcspi200_spi_read(spi, SPI_DATA);
-	*(u8 *)din = tmp_data;
-#endif
+	if (spi->data_merge)
+		*(u32 *)din = atcspi200_spi_read(spi, SPI_DATA);
+	else
+		*(u8 *)din = atcspi200_spi_read(spi, SPI_DATA);
 	return bytes;
 }
 
-static int transfer_data(struct atcspi200_spi *spi, void *rx_buf, void *tx_buf, int num_blks)
+static int transfer_data(struct atcspi200_spi *spi, void *rx_buf, const void *tx_buf, int num_blks)
 {
-#ifdef	CONFIG_SPI_ATCSPI200_DATA_MERGE
-	u32 *dout = (u32 *)tx_buf;
-	u32 *din = (u32 *)rx_buf;
-#else
-	u8 *dout = (u8 *)tx_buf;
-	u8 *din = (u8 *)rx_buf;
-#endif
-	unsigned int event, rx_bytes;
+	unsigned int event;
 	int timeout = spi->timeout;
 	int tx_count = 0, rx_count = 0;
 	int format_val;
@@ -200,8 +191,8 @@ static int transfer_data(struct atcspi200_spi *spi, void *rx_buf, void *tx_buf,
 		event = atcspi200_spi_read(spi, SPI_STATUS);
 		tx_not_full = !(event & ATCSPI200_STATUS_TXFULL_OFFSET);
 		if (tx_not_full && tx_buf && tx_count) {
-			atcspi200_spi_tx(spi, dout);
-			dout += 1;
+			atcspi200_spi_tx(spi, tx_buf);
+			tx_buf = (u8 *)tx_buf + num_byte;
 			tx_count = tx_count - num_byte;
 		}
 
@@ -214,16 +205,16 @@ static int transfer_data(struct atcspi200_spi *spi, void *rx_buf, void *tx_buf,
 		event = atcspi200_spi_read(spi, SPI_STATUS);
 		rx_num = event & ATCSPI200_STATUS_RXNUM_LOWER_MASK;
 		if (rx_num && rx_buf && rx_count) {
-			atcspi200_spi_rx(spi, din, transfer_bytes);
-			din = din + 1;
+			atcspi200_spi_rx(spi, rx_buf, transfer_bytes);
+			rx_buf = (u8 *)rx_buf + num_byte;
 			rx_count = rx_count - num_byte;
 		}
 		num_blks = max(tx_count, rx_count);
 	}
 
 	spi->data_len -= spi->trans_len;
-	spi->din = din;
-	spi->dout = dout;
+	spi->din = rx_buf;
+	spi->dout = tx_buf;
 	return 0;
 
 }
@@ -452,6 +443,16 @@ static int atcspi200_exec_mem_op(struct spi_mem *mem, const struct spi_mem_op *o
 	format_val = atcspi200_spi_read(spi, SPI_TRANSFMT);
 
 	if (op->data.nbytes) {
+		if (IS_ENABLED(CONFIG_SPI_ATCSPI200_DATA_MERGE)) {
+			if ((DATA_BIT == 8) && !(op->data.nbytes % 4)) {
+				format_val |= ATCSPI200_TRANSFMT_DATA_MERGE_MASK;
+				spi->data_merge = true;
+			} else {
+				format_val &= ~ATCSPI200_TRANSFMT_DATA_MERGE_MASK;
+				spi->data_merge = false;
+			}
+			atcspi200_spi_write(spi, SPI_TRANSFMT, format_val);
+		}
 		if (op->data.dir == SPI_MEM_DATA_IN) {
 			spi->din = op->data.buf.in;
 			spi->dout = 0;
diff --git a/include/linux/soc/andes/spi-atcspi200.h b/include/linux/soc/andes/spi-atcspi200.h
index 4fee6c0670a9..cd3330743f50 100644
--- a/include/linux/soc/andes/spi-atcspi200.h
+++ b/include/linux/soc/andes/spi-atcspi200.h
@@ -28,8 +28,10 @@
 #define CHUNK_SIZE		1
 #define SPI_TIMEOUT		0x100000
 #define NSPI_MAX_CS_NUM		1
-#define DATA_LENGTH(x)		((x - 1) << 8)
-#define ADDR_LENGTH(x)		((x - 1) << 16)
+#define DATA_BIT		0x8
+#define DATA_LENGTH(x)		((x - 1) << ATCSPI200_TRANSFMT_DATA_LEN_OFFSET)
+#define ADDR_BIT		0x3
+#define ADDR_LENGTH(x)		((x - 1) << ATCSPI200_TRANSFMT_ADDR_LEN_OFFSET)
 #define DATA_MERGE		(DATA_MERGE_EN << 7)
 #define DMA_TRANSFER_MIN	0x100
 
@@ -112,14 +114,12 @@ struct ts_buf {
 	__be16 data;
 } __packed;
 
-static bool ts_enable;
-
 struct atcspi200_spi;
 struct atcspi200_dma_ops {
 	int (*dma_init)(struct device *dev, struct atcspi200_spi *spi);
 	void (*dma_exit)(struct atcspi200_spi *spi);
-	int (*dma_setup)(struct atcspi200_spi *spi, struct spi_mem_op *op);
-	int (*dma_transfer)(struct atcspi200_spi *spi, struct spi_mem_op *op);
+	int (*dma_setup)(struct atcspi200_spi *spi, const struct spi_mem_op *op);
+	int (*dma_transfer)(struct atcspi200_spi *spi, const struct spi_mem_op *op);
 	void (*dma_stop)(struct atcspi200_spi *spi);
 };
 
@@ -132,13 +132,8 @@ struct atcspi200_spi {
 	size_t			cmd_len;
 	u32			clk_rate;
 	u8			cmd_buf[16];
-#ifdef	CONFIG_SPI_ATCSPI200_DATA_MERGE
-	u32			*din;
-	u32			*dout;
-#else
-	u8			*din;
-	u8			*dout;
-#endif
+	void			*din;
+	const void		*dout;
 	struct ts_buf		*tx_buf;
 	struct ts_buf		*rx_buf;
 	unsigned int		addr;
@@ -149,6 +144,7 @@ struct atcspi200_spi {
 	int			timeout;
 	spinlock_t		lock;
 	struct mutex		mutex_lock;
+	bool			data_merge;
 
 	/* DMA info */
 	struct dma_chan		*txchan;
@@ -158,5 +154,10 @@ struct atcspi200_spi {
 	struct completion	dma_completion;
 };
 
+#ifdef CONFIG_SPI_ATCSPI200_DMA
 void atcspi200_spi_dma_ops_setup(struct atcspi200_spi *spi);
+#else
+inline void atcspi200_spi_dma_ops_setup(struct atcspi200_spi *spi) {}
+#endif
+
 #endif /* !__LINUX_SOC_ANDES_ATCSPI_H */
-- 
2.34.1

